libs/discord/discord.cpp

```diff
@@ -0,0 +1,212 @@
+
+#define CLIENT_ID 882369979406753812
+
+using tDiscordCreate = decltype(&DiscordCreate);
+
+static HMODULE discord_sdk;
+static tDiscordCreate pDiscordCreate = nullptr;
+
+struct Application {
+	IDiscordCore* core;
+	IDiscordUserManager* users;
+	IDiscordActivityManager* activities;
+	DiscordUserId user_id;
+};
+
+
+template< size_t size >
+void Q_strncpyz( char (&dest)[size], const char * src )
+{
+	Q_strncpyz( dest, src, size );
+}
+
+
+
+void OnUserUpdated(void* data)
+{
+	Application* app = (Application*)data;
+	DiscordUser user;
+	app->users->get_current_user(app->users, &user);
+	app->user_id = user.id;
+
+	Com_GGPrint("Discord: Hello {}#{}", user.username, user.discriminator);
+}
+
+void UpdateActivityCallback(void* data, EDiscordResult result)
+{
+	Com_GGPrint("Discord: ActivityCallback: {}", result);
+}
+
+const char* log2str(EDiscordLogLevel level)
+{
+	switch(level)
+	{
+		case DiscordLogLevel_Error: return "ERR ";
+		case DiscordLogLevel_Warn: return "WARN";
+		case DiscordLogLevel_Info: return "INFO";
+		case DiscordLogLevel_Debug: return "DBG ";
+		default: return "?--?";
+	}
+}
+
+void OnLog(void* hook_data, EDiscordLogLevel level, const char* message)
+{
+	Com_GGPrint("Discord: [{}]: {}", log2str(level), message);
+}
+
+static Application app = {};
+static IDiscordUserEvents users_events = {};
+static IDiscordActivityEvents activities_events = {};
+
+
+void InitDiscord()
+{
+	ZoneScoped;
+
+	discord_sdk = LoadLibraryW(L"discord_game_sdk.dll");
+	if (discord_sdk == NULL)
+		return;
+
+	pDiscordCreate = (tDiscordCreate)GetProcAddress(discord_sdk, "DiscordCreate");
+	if (pDiscordCreate == NULL)
+		return;
+
+
+	users_events.on_current_user_update = OnUserUpdated;
+
+	DiscordCreateParams params;
+	DiscordCreateParamsSetDefault(&params);
+	params.client_id = CLIENT_ID;
+	params.flags = DiscordCreateFlags_NoRequireDiscord;	// Fail if discord is not running instead of closing the game
+	params.event_data = &app;
+	params.user_events = &users_events;
+	params.activity_events = &activities_events;
+	EDiscordResult result = pDiscordCreate(DISCORD_VERSION, &params, &app.core);
+	if (result != DiscordResult_Ok) {
+		Com_GGPrint("Discord: not initialized({})", result);
+		// Be sure it's disabled
+		app.core = nullptr;
+	} else {
+		app.core->set_log_hook(app.core, DiscordLogLevel_Debug, nullptr, OnLog);
+
+		app.users = app.core->get_user_manager(app.core);
+		app.activities = app.core->get_activity_manager(app.core);
+	}
+}
+
+
+static connstate_t old_state = CA_UNINITIALIZED;
+static int old_team = -1;
+static int old_match_state = -1;
+static int old_round = -1;
+
+
+void DiscordFrame()
+{
+	ZoneScoped;
+
+	if (app.activities)
+	{
+		if( (cls.state == CA_CONNECTED || cls.state == CA_ACTIVE) && client_gs.module )
+		{
+
+			if (old_state != cls.state ||
+				old_team != cg.predictedPlayerState.real_team ||
+				old_match_state != client_gs.gameState.match_state ||
+				old_round != client_gs.gameState.round_num)
```

Comment:
This whole thing is hacked together a bit, and should probably be a bit stricter to only actually change the status when there is a change that will reflect in other text, since discord is rate-limiting these changes.